handle no_std targets on std builds

onur-ozkan opened the pull request:
This PR unifies the `Step::run_make` logic and improves it by skipping std specific crates for no_std targets. In addition, since we now handle library crates properly, bootstrap is capable of running `x doc library` even for no_std targets as it is able to generate documentation for `alloc` crate from the standard library.

Resolves #128027

cc @ChrisDenton

onur-ozkan committed:
handle no_std targets on std builds

This change unifies the `Step::run_make` logic and improves it by skipping
std specific crates for no_std targets.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
allow running `x doc` on std for no_std targets

Since we now handle library crates properly, there's no need to panic for `no_std`
targets anymore.

`x doc library` now generates documentation for the `alloc` crate from standard library.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10094710212/job/27913176773) [(plain)](https://github.com/rust-lang/rust/commit/0d116b73dc9c7b90bc8c8aa59e27c360a781772a/checks/27913176773/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:3aacb9c90579defe09351ac5e8ee504359f8054da6326ff19038f1b7c90e3cb2aafe33685c6d9b76ee8d2ccbd187ca80c46ab5380485abdd8c0ce7d69cd8d8fd:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.84s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---

---- core::builder::tests::alias_and_path_for_library stdout ----
thread 'core::builder::tests::alias_and_path_for_library' panicked at src/core/builder/tests.rs:204:5:
assertion `left == right` failed
  left: [Std { target: A-A, compiler: Compiler { stage: 0, host: A-A }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }, Std { target: A-A, compiler: Compiler { stage: 1, host: A-A }, crates: ["alloc", "core", "panic_abort", "panic_unwind", "proc_macro", "std", "sysroot", "test", "unwind"], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }]
 right: [Std { target: A-A, compiler: Compiler { stage: 0, host: A-A }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }, Std { target: A-A, compiler: Compiler { stage: 1, host: A-A }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }]

---- core::builder::tests::defaults::build_cross_compile stdout ----
thread 'core::builder::tests::defaults::build_cross_compile' panicked at src/core/builder/tests.rs:270:9:
assertion failed: `(left == right)`
```

</details>

ChrisDenton commented:
This seems to work locally on individual but I am getting ICEs when checking a lot of targets at once. E.g. `./x check std aarch64-kmc-solid_asp3` works on its own but when included with a large number of other targets I get:

```
thread 'rustc' panicked at compiler\rustc_codegen_ssa\src\back\link.rs:86:13:
compiler\rustc_codegen_ssa\src\back\link.rs:86:13: invalid output type `Executable` for target os `aarch64-kmc-solid_asp3`
```

I've not managed to narrowed it down but it could be there's a specific other target that's causing issues when they're used together.

onur-ozkan commented:
> it could be there's a specific other target that's causing issues when they're used together.

Probably.

onur-ozkan committed:
remove the requirement of `Builder` arg in `doc::Std::new` function

`crates` field is handled in the `Step::make_run` just like in any other
`Std` implementation, so we don't need to resolve them in `Std::new`.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
ignore `crates` if running unit tests

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented on src/bootstrap/src/core/build_steps/compile.rs:
```diff
@@ -424,6 +420,28 @@ fn copy_self_contained_objects(
 
+/// Resolves standard library crates for `Std::run_make` for any build kind (like check, build, clippy, etc.).
+pub fn std_crates_for_run_make(run: &RunConfig<'_>) -> Vec<String> {
+    // FIXME: Extend builder tests to cover the `crates` field of `Std` instances.
+    if cfg!(feature = "bootstrap-self-test") {
+        return vec![];
+    }
```
Additional context:

https://github.com/rust-lang/rust/blob/72d73cec61aa8f85901358cd5d386d5dd066fe52/src/bootstrap/src/core/builder/tests.rs#L60-L62

That above and `doc_std` helper macros needs to be improved to mock `crates` field in `Std` with relevant values.

onur-ozkan commented:
@rustbot ready

Mark-Simulacrum commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 47122e3ec14c3d5955cbd120c2d2ce6cd5490fa8 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 47122e3ec14c3d5955cbd120c2d2ce6cd5490fa8 -->
<!-- homu: {"type":"Approved","sha":"47122e3ec14c3d5955cbd120c2d2ce6cd5490fa8","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #125443) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

onur-ozkan commented:
Rebased.

@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit f56d58e4a784b6429e742a411c2759900741894d has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum f56d58e4a784b6429e742a411c2759900741894d -->
<!-- homu: {"type":"Approved","sha":"f56d58e4a784b6429e742a411c2759900741894d","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
